### ISSUE TITLE
Deploy error Could not scan for classes inside "database"

### ISSUE THREAD

**Issue by syui** [cloudfoundry/php-buildpack#293]
buildpack version : v4.3.72

composer php : 7.1.27

src : https://github.com/CachetHQ/Cachet

### error 

```sh
[RuntimeException]
     Could not scan for classes inside "database" which does not appear to be a file nor a folder
```
### setting file

```sh
$ cat composer.json
{
    "name": "cachethq/cachet",
    "description": "An open source status page system, for everyone.",
    "keywords": [
        "laravel",
        "cachet",
        "alt-three",
        "status",
        "page"
    ],
    "type": "project",
    "license": "BSD-3-Clause",
    "authors": [
        {
            "name": "James Brooks",
            "email": "james@alt-three.com"
        },
        {
            "name": "Graham Campbell",
            "email": "graham@alt-three.com"
        },
        {
            "name": "Joseph Cohen",
            "email": "joe@alt-three.com"
        }
    ],
    "replace": {
        "paragonie/random_compat": "*",
        "symfony/polyfill-mbstring": "*"
    },
    "require": {
        "php": "7.1.27",
        "ext-mbstring": "*",
        "ext-openssl": "*",
        "ext-xml": "*",
        "alt-three/badger": "^5.1",
        "alt-three/bus": "^4.1",
        "alt-three/emoji": "^7.0",
        "alt-three/twitter": "^3.1",
        "alt-three/validator": "^4.1",
        "aws/aws-sdk-php": "^3.7",
        "barryvdh/laravel-cors": "^0.11.0",
        "doctrine/dbal": "2.9.*",
        "fideloper/proxy": "^4.0",
        "graham-campbell/binput": "^6.0",
        "graham-campbell/exceptions": "^11.1",
        "graham-campbell/markdown": "^10.2",
        "guzzlehttp/guzzle": "^6.3.3",
        "jenssegers/date": "^3.4",
        "laravel/framework": "5.7.*",
        "laravel/tinker": "^1.0",
        "laravolt/avatar": "^2.1",
        "mccool/laravel-auto-presenter": "^7.1",
        "nexmo/client": "^1.5",
        "pragmarx/google2fa": "^0.7.1",
        "predis/predis": "^1.1",
        "twig/twig": "^2.6"
    },
    "require-dev": {
        "ext-sqlite3": "*",
        "alt-three/testbench": "^4.0",
        "barryvdh/laravel-debugbar": "^3.2",
        "filp/whoops": "^2.3",
        "fzaninotto/faker": "^1.8",
        "graham-campbell/analyzer": "^2.1",
        "graham-campbell/testbench-core": "^3.0",
        "mockery/mockery": "^1.2",
        "phpunit/phpunit": "^7.4",
        "tightenco/mailthief": "^0.3.14"
    },
    "suggest": {
        "ext-apc": "APC Support cache driver."
    },
    "autoload": {
        "classmap": [
            "database"
        ],
        "files": [
            "app/helpers.php"
        ],
        "psr-4": {
            "CachetHQ\\Cachet\\": "app/"
        }
    },
    "autoload-dev": {
        "psr-4": {
            "CachetHQ\\Tests\\Cachet\\": "tests/"
        }
    },
    "scripts": {
        "post-root-package-install": [
          "@php -r \"file_exists('.env') || copy('.env.example', '.env');\""
        ],
        "post-create-project-cmd": [
          "@php artisan key:generate"
        ],
        "post-autoload-dump": [
          "Illuminate\\Foundation\\ComposerScripts::postAutoloadDump",
          "@php artisan package:discover"
        ]
    },
    "config": {
        "platform": {
            "php": "7.1.27"
        },
        "preferred-install": "dist",
        "sort-packages": true,
        "optimize-autoloader": true
    },
    "extra": {
        "branch-alias": {
            "dev-master": "2.4-dev"
        }
    },
    "minimum-stability": "dev",
    "prefer-stable": true
}

$ cat .bc-config/options.json
{
    "COMPOSER_INSTALL_OPTIONS_DEBUG": [ "--no-dev", "-o"],
    "PHP_VERSION": "{PHP_71_LATEST}",
    "COMPOSER_VENDOR_DIR": "vendor",
    "WEBDIR": "public",
    "PHP_EXTENSIONS": ["bz2", "pdo", "zlib", "curl", "mcrypt", "mbstring", "mysqli", "gd", "zip", "pdo_mysql", "pdo_sqlite","iconv",  "openssl" ]
}

$ cat manifest.yml
applications:
- path: .
  name: APP_NAME
  memory: 256M
  buildpack: https://github.com/cloudfoundry/php-buildpack
  command: null
  routes:
  - route: APP_NAME.mybluemix.net

$ cp .env.example .env 
$ vim .gitignore
- .env

$ composer install --no-dev -o
$ composer dump-autoload
$ php artisan key:generate

$ git add .
$ git commit -m "first"
$ bx cf push cachet
.
.
.
Generating optimized autoload files
[RuntimeException]
     Could not scan for classes inside "database" which does not appear to be a file nor a folder
```

> composer.json

```sh
    "autoload": {
        "classmap": [
            "database"
        ],
```

**Comment by syui** [comment]
```sh
$ mkdir htdocs
$ mv database htdocs/
$ vim composer.json
    "autoload": {
        "classmap": [
            "htdocs/database"
        ],
```

...ok

**Comment by dmikusa-pivotal** [comment]
Glad you were able to figure this out. It sounds like you got hit with the buildpack moving your files into the `htdocs` directory. By default, if your app doesn't have an `htdocs` or `WEBDIR` directory, then the buildpack will create one and move all your files there.

By adding an `htdocs` directory and moving your database folder there, it makes the buildpack not move your files. It's a little confusing though because I see you have `WEBDIR` set to `public`. If you change the `WEBDIR` and that folder exists then the buildpack should not look for `htdocs` anymore.

Would need to see `BP_DEBUG=true` output to know for sure though.

**Comment by syui** [comment]
@dmikusa-pivotal Thank you very much for your quick reply.
I succeeded in deploy with the same soruce. However, it came to fail.

The cause is `php htdocs/artisan package: discover` .

Writing "rm -rf htdocs / bootstrap / cache / *" to composer.json has no effect.

```sh
$ bx cf push $APP_NAME
> .bc-config/option.json
> $ composer install --no-dev -o
> ...ok
> compose.json
> "post-autoload-dump": [
>          "Illuminate\\Foundation\\ComposerScripts::postAutoloadDump",
> ...ok
>	  "rm -rf htdocs/bootstrap/cache/*",
> ...ok
>          "@php htdocs/artisan package:discover"
> ...error : PHP Warning:  require(/tmp/app/htdocs/bootstrap/../vendor/autoload.php): L28
```

> htdocs/bootstrap/autoload.php

```php
require __DIR__.'/../vendor/autoload.php';
```

**Comment by dmikusa-pivotal** [comment]
I think I've run Cachet in the past on PWS without any issues. I'll have to give it another try.

Not sure it'll help, but I would try the following:

1. Simplify your `.bp-config/options.json` file to this.

```
{
    "WEBDIR": "public"
}
```

You don't need extensions here because the buildpack will load them from composer.json. The same goes for the PHP version (unless it's not picking the exact one you want). The composer arguments shouldn't be necessary either. The buildpack will know where to put the vendor directory and that works in most cases. I've only see it need to be adjusted when the project is doing something weird like hard coding a path to vendored dependencies.

HTH

**Comment by syui** [comment]
Indeed in the past cachet worked with ibm cloud.

**Comment by syui** [comment]
Should I "composer.lock" add to ".cfignore"?

**Comment by syui** [comment]
1. I simplified `.bc-config/option.json` (WEBDIR only).

2. I reverted htdocs folder.

But I get the same error.

```sh
$ bx cf push $APP_NAME
   Generating optimized autoload files
   > Illuminate\Foundation\ComposerScripts::postAutoloadDump
   > @php artisan package:discover
   PHP Warning:  require(/tmp/app/bootstrap/../vendor/autoload.php): failed to open stream: No such file or directory in /tmp/app/bootstrap/autoload.php on line 28
   PHP Fatal error:  require(): Failed opening required '/tmp/app/bootstrap/../vendor/autoload.php' (include_path='../lib/php:/tmp/app/lib') in /tmp/app/bootstrap/autoload.php on line 28
   Script @php artisan package:discover handling the post-autoload-dump event returned with error code 255
```

**Comment by syui** [comment]
I removed composer.lock.

Necessary add : composer.json

```json
    "require": {
	"ext-pdo":"*",
	"ext-gd":"*",
	"ext-fileinfo":"*"
}
```

**Comment by syui** [comment]
But I get the same error.

**Comment by syui** [comment]
Resolved with cachet commit 04b1cb54945368644b1e3a8681621e75a777b94d .
